### PR TITLE
Enhance Association Request Timeout (1284)

### LINKED ISSUE #1284: Association Request Timout Question/Feature
I am labeling this as a feature request because it just doesn't seem like a bug to me. 

Currently in DicomClient.ClientOptions I can set the AssociationRequestTimeoutInMs.  When using SendAsync with no cancellation token and the timeout expires the connection closes, but another connection/association is attempted because of the requests in the send queue.  This just loops forever.

I could use a the SendAsync with a cancellation token that has a set timout and that does work, but the problem is that if this is set to low the client will close the connection waiting for a move response.  This is really troublsome if there are no pending move responses sent back, only the final move response.  

**What is, according to you, missing from fo-dicom? Please describe.**
If the Association Timeout actually closes the connection and flushes the pending requests queue.

**Describe the solution you'd like**
In DicomClientRequestAssociationState change

            if (winner == associationRequestTimesOut)
            {
                _dicomClient.Logger.Warn($"[{this}] Association request timeout");
                return await _dicomClient.TransitionToCompletedState(_initialisationParameters, cancellation).ConfigureAwait(false);
            }

to something like this

            if (winner == associationRequestTimesOut)
            {
                _dicomClient.Logger.Warn($"[{this}] Association request timeout");
                var exception = new DicomAssociationRequestTimedoutException();  // this is just an example it can be whatever
                return await _dicomClient.TransitionToCompletedWithErrorState(_initialisationParameters, exception, cancellation).ConfigureAwait(false);
            }

Something like this still allows the cancellation token timeout to be much higher and not having to wait for the longer timeout in the association request.

I have done a bit of testing with this and do not really see any issues, although I know there is extensive work currently going on with the DicomClient.  Thoughts? 
 
If this is acceptable I can submit a Pull Request.

### PR DESCRIPTION
**Pull request by BobSter3000**
Update associatonRequestTimesout state in  DicomClientAssociationState

Fixes #1284.

#### Checklist
- [X] The pull request branch is in sync with latest commit on the *fo-dicom/development* branch
- [x] I have updated API documentation
- [x] I have included unit tests
- [X] I have updated the change log
- [X] I am listed in the CONTRIBUTORS file

#### Changes proposed in this pull request:
- Updates DicomClient so that the state transitions to TransitionToCompletedWithErrorState instead of TransitionToCompletedState when the association times out a few times (configurable via setting).
- Add new Exception for TransitionToCompletedWithErrorState

### CODE REVIEW

**Comment by amoerie** [comment]
Looks like a unit test is failing now, in the v5-net-framework build.
Since the net-core-21 and net-core-31 builds are succeeding, we can assume this bug doesn't always occur but is circumstantial.

This is the exception:

```
[xUnit.net 00:00:16.80]     FellowOakDicom.Tests.Network.AsyncDicomCEchoProviderTests.OnCEchoRequestAsync_ShouldRespond [FAIL]
  Failed FellowOakDicom.Tests.Network.AsyncDicomCEchoProviderTests.OnCEchoRequestAsync_ShouldRespond [12 s]
  Error Message:
   FellowOakDicom.Network.DicomAssociationRequestTimedoutException : Association Request timed out waiting for response
  Stack Trace:
     at FellowOakDicom.Network.Client.States.DicomClientCompletedState.<GetNextStateAsync>d__8.MoveNext() in D:\a\fo-dicom\fo-dicom\FO-DICOM.Core\Network\Client\States\DicomClientCompletedState.cs:line 127
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at FellowOakDicom.Network.Client.DicomClient.<Transition>d__91.MoveNext() in D:\a\fo-dicom\fo-dicom\FO-DICOM.Core\Network\Client\DicomClient.cs:line 267
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at FellowOakDicom.Network.Client.States.DicomClientConnectState.<GetNextStateAsync>d__5.MoveNext() in D:\a\fo-dicom\fo-dicom\FO-DICOM.Core\Network\Client\States\DicomClientConnectState.cs:line 82
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at FellowOakDicom.Network.Client.DicomClient.<Transition>d__91.MoveNext() in D:\a\fo-dicom\fo-dicom\FO-DICOM.Core\Network\Client\DicomClient.cs:line 267
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at FellowOakDicom.Network.Client.States.DicomClientIdleState.<SendAsync>d__5.MoveNext() in D:\a\fo-dicom\fo-dicom\FO-DICOM.Core\Network\Client\States\DicomClientIdleState.cs:line 58
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at FellowOakDicom.Network.Client.DicomClient.<SendAsync>d__110.MoveNext() in D:\a\fo-dicom\fo-dicom\FO-DICOM.Core\Network\Client\DicomClient.cs:line 377
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at FellowOakDicom.Tests.Network.AsyncDicomCEchoProviderTests.<OnCEchoRequestAsync_ShouldRespond>d__2.MoveNext() in D:\a\fo-dicom\fo-dicom\Tests\FO-DICOM.Tests\Network\AsyncDicomCEchoProviderTests.cs:line 51
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
```

There's a few things at play here:
- the default association request timeout is 5 seconds
- we are not guaranteed to have a very performant CI server
- even if we have a performant CI server, it can be busy with another test in parallel that is very demanding

Previously, if an association request times out, the DicomClient would cycle through its cleanup states and eventually retry.
You could argue this improves the stability of the DicomClient. 
On the other hand, if it keeps timing out, it's stuck in an endless loop, which is also not good.

I'm now starting to doubt whether this is an improvement. Wouldn't our efforts better be focused on adding an "escape hatch" if the association request keeps timing out? For example, we could expose a new event `DicomClient.AssociationRequestTimedOut`
This would allow consumers to count the number of times this happens and then set a CancellationTokenSource to canceled, which would cancel `SendAsync`. This is just one option.

What do you think @BobSter3000 ? I'm open to suggestions.

**Comment by gofal** [comment]
Agree. Or a property MaxRetryCount, which is very common in many libraries.

**Comment by amoerie** [comment]
I like the idea of a MaxRetryCount, but then we do need to think what should happen when a MaxRetryCount = 3 and a new request is added after attempt 2. Does the retry count reset or not? 
If the next attempt fails, is that then failure number 3 (for the already queued requests) or failure number 1 (for the newly added request)…

I think resetting the counter makes the most sense. Also, I believe that the number of people who add requests while already sending will be a minority.

**Comment by BobSter3000** [comment]
I lke the idea of the event and letting the consumer decide how to handle it with an event,  Just like the RequestTimed Event.  I will see what I can come up with.

**Comment by BobSter3000** [comment]
@amoerie First attempt at implemeting an event that fires when the Assoication Request times out.  There is also a max retry count for the association requests.   If the retry count exceeds the max retry count the exception is thrown.  I am not sure if this is OK or not.
Default Max Retry Count = 3.
Association Request sent
Association Request not responded within timeout Event Fired and Retry Count updated retry
Once Max Retry count is exceeded TransitionToCompletedWithErrorState with the timeout exception

File: FO-DICOM.Core/Network/Client/DicomClientOptions.cs [hunk 1]
```diff
@@ -27,6 +27,11 @@ public class DicomClientOptions
         /// </summary>
         public int? MaximumNumberOfRequestsPerAssociation { get; set; } = null;
 
+        /// <summary>
+        /// The maximum number of Association Request Attempts after AssociationRequestTimeoutInMs expires
```
Comment (amoerie) [inline FO-DICOM.Core/Network/Client/DicomClientOptions.cs:31]: I would rephrase this to "The maximum number of times an association request will be sent before giving up (in the event of a timeout, the request will be retried),  use <see cref="AssociationRequestTimeoutInMs" /> to configure the timeout"

File: FO-DICOM.Core/Network/Client/States/DicomClientRequestAssociationState.cs [hunk 2]
```diff
@@ -204,7 +204,17 @@ public override async Task<IDicomClientState> GetNextStateAsync(DicomClientCance
             if (winner == associationRequestTimesOut)
             {
                 _dicomClient.Logger.Warn($"[{this}] Association request timeout");
-                return await _dicomClient.TransitionToCompletedState(_initialisationParameters, cancellation).ConfigureAwait(false);
+                _dicomClient.AssociationRequestRetryCount++;
+                _dicomClient.NotifyAssociationRequestTimedOut(new EventArguments.AssociationRequestTimedOutEventArgs(_dicomClient.ClientOptions.AssociationRequestTimeoutInMs, _dicomClient.AssociationRequestRetryCount));
+                if (_dicomClient.AssociationRequestRetryCount > _dicomClient.ClientOptions.MaximumNumberOfAssociationRequestAttempts)
+                {
+                    var exception = new DicomAssociationRequestTimedoutException();
```
Comment (amoerie) [inline FO-DICOM.Core/Network/Client/States/DicomClientRequestAssociationState.cs:211]: I think you have to reset the AssociationRequestRetryCount here. Otherwise, in future SendAsync invocations, the request will not be retried.

File: FO-DICOM.Core/Network/Client/EventArguments/AssociationRequestTimedOutEventArgs.cs [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+
+namespace FellowOakDicom.Network.Client.EventArguments
+{
+    public class AssociationRequestTimedOutEventArgs : EventArgs
+    {
+
+        /// <summary>
+        /// Gets the timeout that was exceded
+        /// </summary>
+        public int Timeout { get; }
```
Comment (amoerie) [inline FO-DICOM.Core/Network/Client/EventArguments/AssociationRequestTimedOutEventArgs.cs:14]: Please either use a TimeSpan or rename the property to TimeoutInMs to make it explicit what the unit of time is.

**Review by amoerie — CHANGES_REQUESTED** [review]
The code changes look good, but I think we should add two missing unit tests:

1. One new test that verifies this new exception is thrown after 3 timeouts. 
2. One new test that verifies association requests are always tried 3 times. 

If you would like some assistance with writing the tests, please let me know. I can give pointers or help you set something up. If you want to take a stab at it, you can find a lot of scenarios in the DicomClientTests and DicomClientTimeoutTests.

P.S. The code coverage check is failing, but it will succeed if you pull in the latest changes from development. We only just enabled code coverage, and the default settings do not allow ANY decrease in coverage. This is very harsh. I changed the config to allow a decrease of 0.1% (which is already more considerable than it sounds);

**Comment by BobSter3000** [comment]
@amoerie I will get started on these changes.  Thank you.

File: FO-DICOM.Core/Network/Client/States/DicomClientRequestAssociationState.cs (hunk 2 above)
Comment (amoerie) [reply FO-DICOM.Core/Network/Client/States/DicomClientRequestAssociationState.cs:211]: You will also have to reset this count if the association request does NOT timeout (accept or reject)

**Comment by BobSter3000** [comment]
@amoerie I am having a problem trying to setup the second unit test requested.
2. One new test that verifies association requests are always tried 3 times.
I just do not see a way to see when the assocication is being requested again.  Maybe one of the state changed events?
The other unit test covers the exception and verifies the retry count.

**Comment by amoerie** [comment]
I can help you out with that but please allow me some time. Thanks for the effort so far!

**Comment by BobSter3000** [comment]
@amoerie Thanks again for all the help on this.

**Comment by amoerie** [comment]
@BobSter3000 I added two tests that verify the counter is reset when an association is accepted or rejected. These confirm that, when a timeout occurs, the same DicomClient can be reused again and will again try 2 or 3 times to associate.

You might notice that I also made some changes in seemingly unrelated parts. One is that the DicomClientConnectState was erroneously catching exception from the transition to the association request state, which since this PR starts throwing the timeout exception. This was wrong, and I moved the TransitionToAssociationRequestState outside the try block. The effect of the original code was that the completed with error state was entered multiple times. This caught my eye in the logs, and is now fixed.

Furthermore, and this was really subtle, I could see that the DicomClient connection cleanup would complete before the listener actually stops. I traced this back to the usage of TaskFactory.StartNew, which doesn't support await lambdas that return a Task. The result was that the Listener always completed much too quickly, causing confusing timings in the DicomClient logs. The solution was extremely simple: use Task.Run instead of TaskFactory.StartNew. We have lost the ability now to pass the flag "LongRunning", which hints at the TaskScheduler that a new dedicated thread should be launched, while Task.Run will borrow a thread from the threadpool. However, after reading up on some of the internal heuristics of the ThreadPool, it will also spin up extra threads if it sees some borrowed threads are long running, so ultimately the result will be the same as far as I can tell.

Sorry for "hijacking" your pull request to include these fixes.

**Review by amoerie — APPROVED** [review]

**Comment by BobSter3000** [comment]
@amoerie I really do appreciate all the.  Thanks again!

**Review by mrbean-bremen — APPROVED** [review]
LGTM

**Review by gofal — APPROVED** [review]

**Comment by amoerie** [comment]
Merged! 🎉

Thanks for contributing @BobSter3000 !
